py/server/deephaven/experimental/sql.py

```diff
@@ -0,0 +1,64 @@
+#
+# Copyright (c) 2016-2023 Deephaven Data Labs and Patent Pending
+#
+import contextlib
+import inspect
+import jpy
+from typing import Dict, Union, Mapping, Optional, Any
+
+from deephaven import DHError
+from deephaven.table import Table, _j_py_script_session
+
+_JSql = jpy.get_type("io.deephaven.engine.sql.Sql")
+
+
+@contextlib.contextmanager
+def _scope(globals: Mapping[str, Any], locals: Mapping[str, Any]):
+    j_py_script_session = _j_py_script_session()
+    # Can do `globals | locals` in Python 3.9+
+    j_py_script_session.pushScope({**globals, **locals})
+    try:
+        yield
+    finally:
+        j_py_script_session.popScope()
+
+
+def eval(
+    sql: str,
+    dry_run: bool = False,
+    globals: Optional[Mapping[str, Any]] = None,
+    locals: Optional[Mapping[str, Any]] = None,
```

Comment:
I am not clear about the significance of separating the scope into globals and locals. Would something like 'extra_scope' be simpler, unless we do modify the current globals and/or locals?

Reply:
This is meant to more most closely match with the scoping model that Python itself uses: https://docs.python.org/3/library/functions.html#exec. I think for 90+% of use cases, the user won't be setting any scoping parameters (so, functionally, it's just as simple to have the defaults for `globals`/`locals` as it is to have the defaults for a single `extra_scope`).

Along this vein, I think there is good potential in the future to re-implement PythonScopeJpyImpl in this fashion as well to keep `globals` / `locals` dicts instead of single `dict` (and potentially, to have the majority of the logic inside python itself to save back and forth boundary crossings); this would potentially allow for python scoping in java to be more efficient (wouldn't need to create a new `dict` each time).

While `extra_scope` would be "simpler" for some use cases, I think there may be advanced use cases that could potentially take advantage of the distinction in the future:

```sql
SELECT * FROM locals.my_table, globals.my_table WHERE locals.my_table.id = globals.my_table.id
```

Reply:
```python
def some_advanced_eval(globals=None, locals=None):
  return eval("...", globals=globals, locals=locals)
```

```sql
SELECT custom_python("some_advanced_eval(globals(), locals())")
```

Reply:
It's not necessarily meant to be "clean" - it's meant to be possible to do advanced things. For example,

`org.jpy.PyLib#executeCode(String code, int start, Object globals, Object locals)`

`org.jpy.PyLib#executeScript(String file, int start, Object globals, Object locals)`

both take globals / locals instead of a single dict; it more closely matches how c python operates, https://docs.python.org/3/c-api/veryhigh.html#c.PyEval_EvalCode

Reply:
If we think passing in globals / locals via parameters is too speculative right now, I might prefer to get rid of it entirely:

```python
def eval(
    sql: str,
    dry_run: bool = False,
) -> Union[Table, jpy.JType]:
...
```

It's not unreasonable to think that this sort of functionality may be more general in the future, and it's up to the user to call it:

```python
from deephaven.experimental import sql, custom_scope

my_globals = ...
my_locals = ...

with custom_scope(my_globals, my_locals):
  my_table = sql.eval("...")
  
# as opposed to:
# my_table = sql.eval("...", globals=my_globals, locals=my_locals)

Reply:
On the python table operations side, there were a lot of discussions between @jmao-denver , @rcaudy, and me about how to express the scoping.  Some of the ones that we decided not to follow look like what is currently implemented here.  As a result, I'm for removing the parameters so that normal and SQL behave the same.  If we get some sort of use case that needs further functionality to explicitly define the values, we will need to make the same change in both places.